desc/protoparse/ast/values.go

```diff
@@ -0,0 +1,416 @@
+	AsFloat() float64
+}
+
+var _ FloatValueNode = (*FloatLiteralNode)(nil)
+var _ FloatValueNode = (*SpecialFloatLiteralNode)(nil)
+var _ FloatValueNode = (*UintLiteralNode)(nil)
+
+type FloatLiteralNode struct {
+	terminalNode
+	Val float64
+}
+
+func NewFloatLiteralNode(val float64, info TokenInfo) *FloatLiteralNode {
+	return &FloatLiteralNode{
+		terminalNode: terminalNode{
+			posRange: info.PosRange,
+			leading:  info.LeadingComments,
+			trailing: info.TrailingComments,
+		},
+		Val: val,
+	}
+}
+
+func (n *FloatLiteralNode) Value() interface{} {
+	return n.AsFloat()
+}
+
+func (n *FloatLiteralNode) AsFloat() float64 {
+	return n.Val
+}
+
+type SpecialFloatLiteralNode struct {
+	*KeywordNode
+	Val float64
+}
+
+func NewSpecialFloatLiteralNode(name *KeywordNode) *SpecialFloatLiteralNode {
+	var f float64
+	if name.Val == "inf" {
+		f = math.Inf(1)
+	} else {
+		f = math.NaN()
+	}
+	return &SpecialFloatLiteralNode{
+		KeywordNode: name,
+		Val:         f,
+	}
+}
+
+func (n *SpecialFloatLiteralNode) Value() interface{} {
+	return n.AsFloat()
+}
+
+func (n *SpecialFloatLiteralNode) AsFloat() float64 {
+	return n.Val
+}
+
+type SignedFloatLiteralNode struct {
+	compositeNode
+	Sign  *RuneNode
+	Float FloatValueNode
+	Val   float64
+}
+
+func NewSignedFloatLiteralNode(sign *RuneNode, f *FloatLiteralNode) *SignedFloatLiteralNode {
+	children := []Node{sign, f}
+	val := f.Val
+	if sign.Rune == '-' {
+		val = -f.Val
+	}
+	return &SignedFloatLiteralNode{
+		compositeNode: compositeNode{
+			children: children,
+		},
+		Sign:  sign,
+		Float: f,
+		Val:   val,
+	}
+}
+
+func (n *SignedFloatLiteralNode) Value() interface{} {
+	return n.Val
+}
+
+type BoolLiteralNode struct {
+	*KeywordNode
+	Val bool
+}
+
+func NewBoolLiteralNode(name *KeywordNode) *BoolLiteralNode {
+	return &BoolLiteralNode{
+		KeywordNode: name,
+		Val:         name.Val == "true",
+	}
+}
+
+func (n *BoolLiteralNode) Value() interface{} {
+	return n.Val
+}
+
+type SliceLiteralNode struct {
+	compositeNode
+	OpenBracket  *RuneNode
+	Elements     []ValueNode
+	Commas       []*RuneNode
+	CloseBracket *RuneNode
+}
+
+func NewSliceLiteralNode(open *RuneNode, vals []ValueNode, commas []*RuneNode, close *RuneNode) *SliceLiteralNode {
+	children := make([]Node, len(vals)*2 + 1)
+	children = append(children, open)
+	for i, val := range vals {
+		if i > 0 {
+			children = append(children, commas[i-1])
+		}
+		children = append(children, val)
+	}
+	children = append(children, close)
+
+	return &SliceLiteralNode{
+		compositeNode: compositeNode{
+			children: children,
+		},
+		OpenBracket:  open,
+		Elements:     vals,
+		Commas:       commas,
+		CloseBracket: close,
+	}
+}
+
+func (n *SliceLiteralNode) Value() interface{} {
+	return n.Elements
+}
+
+type AggregateLiteralNode struct {
+	compositeNode
+	Open     *RuneNode
+	Elements []*AggregateEntryNode
+	Close    *RuneNode
+}
+
+func NewAggregateLiteralNode(open *RuneNode, vals []*AggregateEntryNode, close *RuneNode) *AggregateLiteralNode {
```

Comment:
What's an example of an aggregate value? I'm assuming these could be used for uncommon use cases like custom options?

Reply:
Yes, message literals in custom options. I haven't finished adding doc comments and examples to all types, but I will. Here's an example:

```proto
option (grpc.gateway.protoc_gen_swagger.options.openapiv2_operation) = {
            summary: "Get User"
            description: "Get data about users who have been identified via the [identify](/identify) Browser API function."
            tags: "Users"
        };
```
The message literal, starting with `{` on the first line and ending with `}` on the last, is the aggregate literal.

The name "aggregate" was used in the current `ast.go`, but I can see a better name is warranted so this makes more sense, like `MessageLiteralNode.`

Reply:
Renamed to `MessageLiteralNode`.